lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -0,0 +1,429 @@
+    bool Condition = boolcast(what);                             \
+    print_assert(Condition, #what, message, __FILE__, __LINE__); \
+  } else {                                                       \
+    revng_check(what, message);                                  \
+  }
+
+// clang-format off
+template<typename T>
+using PlainStorage = std::conditional_t<anyOf<T, char *, const char *>(),
+                                        std::vector<char>,
+                                        T>;
+template<typename T>
+using ArgStorage = std::conditional_t<isList<T>(),
+                                      std::vector<std::remove_pointer_t<T>>,
+                                      PlainStorage<T>>;
+// clang-format on
+
+static llvm::ExitOnError AbortOnError;
+
+namespace utils {
+// Utility functions for decoding arguments
+
+// Put a nullptr at the end of the array and poison it, this will make it so
+// out-of-bounds errors are caught quickly
+template<typename T>
+inline void poisonTail(std::vector<T *> &Vector) {
+  Vector.push_back(nullptr);
+  ASAN_POISON_MEMORY_REGION(Vector.back(), sizeof(T *));
+}
+
+inline std::vector<const char *> toStringList(ArgumentRef Argument) {
+  std::vector<const char *> Result;
+  std::vector<llvm::StringRef> ArgList = Argument.toStringList();
+  for (llvm::StringRef Arg : ArgList) {
+    Result.push_back(Arg.data());
+  }
+  poisonTail(Result);
+  return Result;
+}
+
+} // namespace utils
+
+static std::vector<string> TemporaryDirectories;
+static void cleanTemporaryDirectories(void *) {
+  for (auto &Directory : TemporaryDirectories) {
+    std::filesystem::remove_all(Directory);
+  }
+}
+
+class RunnerContext {
+private:
+  llvm::StringMap<uintptr_t> Pointers;
+
+public:
+  const revng::tracing::RunTraceOptions Options;
+
+public:
+  RunnerContext(const revng::tracing::RunTraceOptions Options) :
+    Options(Options) {}
+  ~RunnerContext() = default;
+  RunnerContext(const RunnerContext &Other) = delete;
+  RunnerContext(RunnerContext &&Other) = delete;
+  RunnerContext &operator=(const RunnerContext &Other) = delete;
+  RunnerContext &operator=(RunnerContext &&Other) = delete;
+
+  const std::string getTemporaryDirectory() {
+    llvm::SmallString<128> TempOutput;
+    llvm::sys::fs::createUniqueDirectory("revng-run-trace", TempOutput);
+    std::string Output = TempOutput.str().str();
+    if (!Options.KeepTemporaries) {
+      TemporaryDirectories.push_back(Output);
+    } else {
+      std::cout << "Creating temporary directory: " << Output << "\n";
+      std::cout.flush();
+    }
+    return Output;
+  }
+
+  template<typename T>
+  void storePointer(const llvm::StringRef Name, T *Pointer) {
+    if (Name == NullPointer) {
+      return;
+    }
+    Pointers[Name] = reinterpret_cast<uintptr_t>(Pointer);
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_pointer_v<T>
+  T getPointer(const llvm::StringRef Name, bool Invalidate) {
```

Comment:
A method called `getPointer` should not have an argument `Invalidate`.
Make a distinct method.

Reply:
Split methods